GraphQL: node set dependencies, full node graph

MarkusHorstmann opened the pull request:
This PR uses a NodeSet model (from the CESMII Profile Designer) and the widely used GraphQL "Hot Chocolate" library to expose a fully resolved object graph of all nodes in the nodesets. It is deployed in the  [CESMII Stage Cloud Library ](https://cloudlib-back-stage.azurewebsites.net/graphiql) for easier evaluation.

Open issues:
- Compatibility with current graphql schema
>> Done.
- Use the CESMII NuGets or clone the source (https://github.com/cesmii/CESMII-NodeSet-Utilities)? 
>> Decided to use the NuGet packages.
- Should invalid models continue to be allowed? (Currently there are 5 such models)
>> Models are allowed but marked as having validation errors.

Try:
```graphql
# Interfaces in all nodesets
query MyQuery {
  interfaces {
    nodes {
      browseName
    }
  }
}
```
```graphql
# DeviceType and its first 3 super  types
query MyQuery {
  objectTypes(nodeId: "nsu=http://opcfoundation.org/UA/DI/;i=1002") {
    nodes {
      browseName nodeId
      properties {
        browseName nodeId
      }
      superType {
        browseName nodeId
        superType {
          browseName nodeId
          superType {
            browseName nodeId
          }
        }
      }
    }
  }
}
```
```graphql
# Object types that contain "Vendor" or "Server" in their browseName
query MyQuery {
  objectTypes(where: {or: [{browseName: {contains: "Vendor"}}, {browseName: {contains: "Server"}}]}) {
    nodes {
      browseName
    }
  }
}
```
```graphql
# Node sets, their required models, and the actual models/versions found in the cloud library
query MyQuery {
  nodeSets {
    nodes {
      modelUri
      publicationDate
      identifier
      requiredModels {
        modelUri
        publicationDate
        version
        model {
          modelUri
          publicationDate
          version
          identifier
        }
      }
    }
  }
}
```

BoBiene commented:
as requested in #106 we need API stability. 
There is a released product using the API (https://www.opc-router.com/opc-router-4-27-update/)

BoBiene reviewed (dismissed):
Compatibility with current graphql schema

BoBiene commented:
@MarkusHorstmann is the https://cloudlib-back-stage.azurewebsites.net/graphiql running the same version as in this pr?

MarkusHorstmann commented:
> @MarkusHorstmann is the https://cloudlib-back-stage.azurewebsites.net/graphiql running the same version as in this pr?

Yes, with some minor changes, but not in the graphql or data areas.

> @MarkusHorstmann is the https://cloudlib-back-stage.azurewebsites.net/graphiql running the same version as in this pr?

Yes, essentially. Identical in graphql and data index: https://github.com/cesmii/CESMII-CloudLibrary/compare/opcContrib/mhdev...cesmii:cesmii/develop

MarkusHorstmann commented:
Fixes #96  and starts addressing #43.

malessainray commented:
> > @MarkusHorstmann is the https://cloudlib-back-stage.azurewebsites.net/graphiql running the same version as in this pr?
> 
> Yes, essentially. Identical in graphql and data index: [cesmii/CESMII-CloudLibrary@opcContrib/mhdev...cesmii:cesmii/develop](https://github.com/cesmii/CESMII-CloudLibrary/compare/opcContrib/mhdev...cesmii:cesmii/develop)

Sadly I cannot login or register there for the live of me. Please get in touch with me so that we can perform tests against the staging deployment.

malessainray commented:
These changes break our client side implementation in shipping product of dependency resolution for nodesets. With the instance hosted at https://uacloudlibrary.opcfoundation.org/ it works fine. I will follow up with more details on what specifically changed.

malessainray commented:
The from the method `Opc.Ua.CloudLib.Client.UACloudLibClient.GetNamespacesAsync()` returned tuple values for the identifiers all have a leading space. This in turn results in failing to download the nodeset using the exact identifier as in no place any string trimming is performed. The stable instance does not return those leading spaces.

MarkusHorstmann commented:
> The from the method `Opc.Ua.CloudLib.Client.UACloudLibClient.GetNamespacesAsync()` returned tuple values for the identifiers all have a leading space. This in turn results in failing to download the nodeset using the exact identifier as in no place any string trimming is performed. The stable instance does not return those leading spaces.

Fixed and added a test.

MarkusHorstmann commented:
> @MarkusHorstmann when will the latest changes in this pullrequest be deployed to the staging instance for integration testing?

It is now deployed at https://cloudlib-back-stage.azurewebsites.net/.

malessainray commented:
I could not spot incompatibilities with our released product.

malessainray commented on SampleConsoleClient/Program.cs:
```diff
@@ -241,10 +309,54 @@ private static async Task TestClientLibraryAsync(string[] args)
+                    foreach (var nodeSet in nodeSets)
+                    {
+                        Console.WriteLine($"Dependencies for {nodeSet.Identifier} {nodeSet.NamespaceUri} {nodeSet.PublicationDate} ({nodeSet.Version}):");
+                        foreach (var requiredNodeSet in nodeSets[0].RequiredModels)
```
```suggestion
                        foreach (var requiredNodeSet in nodeSet.RequiredModels)
```

barnstee commented on UACloudLibraryServer/AWSFileStorage.cs:
```diff
@@ -186,5 +186,34 @@ public async Task<string> DownloadFileAsync(string name, CancellationToken cance
             }
         }
+        /// <summary>
+        /// Find a file based on a unique name
```
comment

MarkusHorstmann replied:
Fixed

barnstee commented on UACloudLibraryServer/Controllers/InfoModelController.cs:
```diff
@@ -54,12 +54,16 @@ public class InfoModelController : ControllerBase
+        private readonly NodeSetModelIndexerFactory _nodeSetIndexerFactory;
 
-        public InfoModelController(IFileStorage storage, IDatabase database, ILoggerFactory logger)
+        public InfoModelController(IFileStorage storage, IDatabase database, ILoggerFactory logger, NodeSetModelIndexer indexer, NodeSetModelIndexerFactory nodeSetIndexerFactory, AppDbContext dbContext)
```
dnContext not used

MarkusHorstmann replied:
Fixed
